### PR TITLE
Add more input validation ;; WLFE-32

### PR DESCRIPTION
**Pull request by ofrogon**
See issue #11

### CODE REVIEW

File: packages/create-workleap/src/bin.ts [hunk 1]
```diff
@@ -12,11 +12,18 @@ let outputDirectory = process.argv[2];
 
 intro(colors.gray(`${packageJson.name} - v${packageJson.version}`));
 
+const invalidPathCharactersRegex = /[:*?"<>|]/;
```
Comment (patricklafrance) [inline packages/create-workleap/src/bin.ts:15]: Constants should start with an uppercase caracter.

File: packages/foundry/src/create.ts [hunk 2]
```diff
@@ -7,11 +7,21 @@ const BaseRepositoryAddress = "workleap/wl-foundry-cli/templates";
 
 type TemplateId = "host-application" | "remote-module" | "static-module";
 
+const validNpmPackageNameRegex = /^(?:@[a-z0-9-*~][a-z0-9-*._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/;
```
Comment (patricklafrance) [inline packages/foundry/src/create.ts:10]: Constants should start with an uppercase caracter.

Comment (patricklafrance) [reply packages/foundry/src/create.ts:10]: I thought the plan was to only validate inputs in the client (create-workleap) project?

File: packages/foundry/src/create.ts [hunk 3]
```diff
@@ -7,11 +7,21 @@ const BaseRepositoryAddress = "workleap/wl-foundry-cli/templates";
...
+        throw new Error(`Invalid package.json name "${scope}/${name}"`);
+    }
+}
+
 const TemplateGenerators: Record<TemplateId, (outputDirectory: string, options: Record<string, string>) => Promise<void>> = {
     "host-application": async (outputDirectory, options) => {
         const scope = options["packageScope"];
         const name = basename(options["outDir"]);
 
+        validatePackageName(scope, name);
```
Comment (alexasselin008) [inline packages/foundry/src/create.ts:23]: I think its a bit late at this point to do validation, as the command is now running. 

Can we run the validNpmPackageNameRegex on the scope during the initial validation?

**Comment by patricklafrance** [comment]
Sorry, this is an issue that should have spot in the previous PR but there were so many things going on that I missed it!.

For the `host-application` template, you don't need to extract a project name from the output directory as the names will be hardcoded to match a pattern.

You don't need this 👇🏻 

```
        const scope = options["packageScope"];
        const name = basename(options["outDir"]);

        validatePackageName(scope, name);
```

For the `host-application` template the pattern will be:

- {{PACKAGE-SCOPE}}/host
- {PACKAGE-SCOPE}}/shell
- {{PACKAGE-SCOPE}}/components
- {{PACKAGE-SCOPE}}/services
- {{PACKAGE-SCOPE}}/types
- {{PACKAGE-SCOPE}}/events

File: packages/foundry/src/create.ts (hunk 3 above)
Comment (alexasselin008) [reply packages/foundry/src/create.ts:23]: We can validate parts of the name

**Comment by patricklafrance** [comment]
As for the `remote-module` and `static-module` templates, indeed a package name will be needed.

I don't think it should be extracter from the output directory thought, it should be in fact the opposite, the output directory could be defaulted to the package name.

Furthermore, the package name can't be extracted from the output directory because it might requires to contain a scope (and it will not necessary be the {{HOST-SCOPE}}.

So here's an updated version of the `remote-module` template tokens:
- {{HOST-SCOPE}}
- {{PACKAGE-NAME}}

And the client should ask the user for:
- The host application package scope
- A package name (package name could be either `my-package` or with a scope `@myscope/mypackage`)
- An output directory if not already provided by `pnpm create`

Here's an updated version of the `static-module` template tokens:
- {{HOST-SCOPE}}
- {{PACKAGE-NAME}}

And the client should ask the user for:
- The host application package scope
- A package name (package name could be either `my-package` or with a scope `@myscope/mypackage`)
- An output directory if not already provided by `pnpm create`

File: packages/foundry/src/create.ts (hunk 3 above)
Comment (patricklafrance) [reply packages/foundry/src/create.ts:23]: Given that we will ask for the package name is the client (create-workleap) the name could be validated at that time as we do for other arguments.

What do you think @alexasselin008 ?

Comment (alexasselin008) [reply packages/foundry/src/create.ts:23]: Yeah

Comment (ofrogon) [reply packages/foundry/src/create.ts:23]: Perfect, I will oupdate the flow and move that validation

**Comment by ofrogon** [comment]
@patricklafrance, I created the issue #16 about the package your comment. I will go on this next.

File: packages/foundry/src/create.ts (hunk 3 above)
Comment (ofrogon) [reply packages/foundry/src/create.ts:23]: Finally, the new option will be done in the next PR and the package name validation will also be added in this next PR.

**Review by alexasselin008 — APPROVED** [review]

**Pull request merged**